File: codex/crm/android/CrmMobileAndroid/app/src/main/java/com/example/crm/android/ApiClient.java
package com.example.crm.android;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.UUID;

public class ApiClient {
    private static final int TIMEOUT_MS = 15000;

    private String baseUrl;
    private String token;

    public ApiClient(String baseUrl, String token) {
        setBaseUrl(baseUrl);
        this.token = token;
    }

    public void setBaseUrl(String baseUrl) {
        if (baseUrl == null) {
            this.baseUrl = "";
            return;
        }
        String trimmed = baseUrl.trim();
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        if (!trimmed.contains("://")) {
            trimmed = "http://" + trimmed;
        }
        this.baseUrl = trimmed;
    }

    public void setToken(String token) {
        this.token = token;
    }

    public LoginResult login(String username, String password) throws ApiException {
        try {
            JSONObject body = new JSONObject();
            body.put("username", username);
            body.put("password", password);
            JSONObject data = request("POST", "/auth/login", body);
            LoginResult result = new LoginResult();
            result.token = data.optString("token");
            result.username = data.optString("username");
            return result;
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("登录失败", ex);
        }
    }

    public PageResult<Customer> listCustomers(int current, int size, String keyword) throws ApiException {
        try {
            StringBuilder path = new StringBuilder("/customers?current=")
                    .append(current)
                    .append("&size=")
                    .append(size);
            if (keyword != null && !keyword.trim().isEmpty()) {
                path.append("&keyword=").append(URLEncoder.encode(keyword.trim(), "UTF-8"));
            }
            JSONObject data = request("GET", path.toString(), null);
            PageResult<Customer> page = new PageResult<>();
            page.total = data.optInt("total");
            page.size = data.optInt("size");
            page.current = data.optInt("current");
            JSONArray records = data.optJSONArray("records");
            if (records != null) {
                for (int i = 0; i < records.length(); i++) {
                    page.records.add(Customer.fromJson(records.getJSONObject(i)));
                }
            }
            return page;
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("加载客户失败", ex);
        }
    }

    public Customer customerDetail(int id) throws ApiException {
        try {
            return Customer.fromJson(request("GET", "/customers/" + id, null));
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("加载客户详情失败", ex);
        }
    }

    public PageResult<FinanceNews> listFinanceNews(int current, int size, int categoryTag) throws ApiException {
        try {
            String path="/news?current="+current+"&size="+size+(categoryTag < 0 ? "" : "&categoryTag="+categoryTag);
            JSONObject data=request("GET", path, null); PageResult<FinanceNews> page=new PageResult<>();
            page.total=data.optInt("total"); page.current=data.optInt("current"); page.size=data.optInt("size"); JSONArray records=data.optJSONArray("records");
            if(records!=null) for(int i=0;i<records.length();i++) page.records.add(FinanceNews.fromJson(records.getJSONObject(i))); return page;
        } catch(Exception ex) { throw new ApiException("加载资讯失败", ex); }
    }

    public PageResult<StockQuote> listStocks(int current, int size, String keyword) throws ApiException {
        try {
            String path="/stocks?current="+current+"&size="+size+"&sortField=changeRate&sortOrder=descend";
            if(keyword!=null&&!keyword.trim().isEmpty()) path += "&keyword="+URLEncoder.encode(keyword.trim(), "UTF-8");
            JSONObject data=request("GET", path, null); PageResult<StockQuote> page=new PageResult<>();
            page.total=data.optInt("total"); page.current=data.optInt("current"); page.size=data.optInt("size");
            JSONArray records=data.optJSONArray("records");
            if(records!=null) for(int i=0;i<records.length();i++) page.records.add(StockQuote.fromJson(records.getJSONObject(i)));
            return page;
        } catch(Exception ex) { throw new ApiException("加载股票行情失败", ex); }
    }

    public PageResult<StockQuote> stockHistory(String stockCode, int current, int size) throws ApiException {
        try {
            JSONObject data=request("GET", "/stocks/"+stockCode+"/history?current="+current+"&size="+size, null);
            PageResult<StockQuote> page=new PageResult<>(); page.total=data.optInt("total"); page.current=data.optInt("current"); page.size=data.optInt("size");
            JSONArray records=data.optJSONArray("records");
            if(records!=null) for(int i=0;i<records.length();i++) page.records.add(StockQuote.fromJson(records.getJSONObject(i)));
            return page;
        } catch(Exception ex) { throw new ApiException("加载股票历史失败", ex); }
    }

    public PageResult<Fund> listFunds(int current, int size, String keyword) throws ApiException {
        return listFunds(current, size, keyword, null, null, null, null);
    }

    public PageResult<Fund> listFunds(int current, int size, String keyword, String fundType,
                                      Boolean canBuy, String sortField, String sortOrder) throws ApiException {
        try {
            StringBuilder path = new StringBuilder("/funds?current=")
                    .append(current)
                    .append("&size=")
                    .append(size);
            if (keyword != null && !keyword.trim().isEmpty()) {
                path.append("&keyword=").append(URLEncoder.encode(keyword.trim(), "UTF-8"));
            }
            appendQuery(path, "fundType", fundType);
            if (canBuy != null) path.append("&canBuy=").append(canBuy);
            appendQuery(path, "sortField", sortField);
            appendQuery(path, "sortOrder", sortOrder);
            JSONObject data = request("GET", path.toString(), null);
            PageResult<Fund> page = new PageResult<>();
            page.total = data.optInt("total");
            page.size = data.optInt("size");
            page.current = data.optInt("current");
            JSONArray records = data.optJSONArray("records");
            if (records != null) {
                for (int i = 0; i < records.length(); i++) {
                    page.records.add(Fund.fromJson(records.getJSONObject(i)));
                }
            }
            return page;
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("加载产品失败", ex);
        }
    }

    private void appendQuery(StringBuilder path, String name, String value) throws Exception {
        if (value != null && !value.trim().isEmpty()) {
            path.append("&").append(name).append("=")
                    .append(URLEncoder.encode(value.trim(), "UTF-8"));
        }
    }

    public FundDetail fundDetail(String fundCode) throws ApiException {
        try {
            return FundDetail.fromJson(request("GET", "/funds/" + URLEncoder.encode(fundCode, "UTF-8"), null));
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("加载产品详情失败", ex);
        }
    }

    public PageResult<FundNav> listFundNavs(String fundCode, int current, int size) throws ApiException {
        try {
            String path = "/funds/" + URLEncoder.encode(fundCode, "UTF-8")
                    + "/navs?current=" + current
                    + "&size=" + size;
            JSONObject data = request("GET", path, null);
            PageResult<FundNav> page = new PageResult<>();
            page.total = data.optInt("total");
            page.size = data.optInt("size");
            page.current = data.optInt("current");
            JSONArray records = data.optJSONArray("records");
            if (records != null) {
                for (int i = 0; i < records.length(); i++) {
                    page.records.add(FundNav.fromJson(records.getJSONObject(i)));
                }
            }
            return page;
        } catch (ApiException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new ApiException("加载净值走势失败", ex);
        }
    }

    public PageResult<UserFundHolding> listPortfolioHoldings(int current, int size, String keyword) throws ApiException {
        return listPortfolioHoldings(current, size, keyword, "raw", "holdingAmount", "desc");
    }

    public PageResult<UserFundHolding> listPortfolioHoldings(int current, int size, String keyword,
                                                              String scope, String sortField,
                                                              String sortOrder) throws ApiException {
        try {
            StringBuilder path = new StringBuilder("/portfolio/holdings?current=")
                    .append(current)
                    .append("&size=")
                    .append(size)
                    .append("&scope=").append(scope)
                    .append("&sortField=").append(sortField)
                    .append("&sortOrder=").append(sortOrder);
            if (keyword != null && !keyword.trim().isEmpty()) {
                path.append("&keyword=").append(URLEncoder.encode(keyword.trim(), "UTF-8"));
            }
            JSONObject data = request("GET", path.toString(), null);
            PageResult<UserFundHolding> page = new PageResult<>();
            page.total = data.optInt("total");
            page.size = data.optInt("size");
            page.current = data.optInt("current");
            JSONArray records = data.optJSONArray("records");
            if (records != null) {
                for (int i = 0; i < records.length(); i++) {
                    page.records.add(UserFundHolding.fromJson(records.getJSONObject(i)));
                }
            }
            return page;
        } catch (Exception ex) {
            throw new ApiException("加载持仓失败", ex);
        }
    }

    public PortfolioOverview portfolioOverview() throws ApiException {
        try {
            return PortfolioOverview.fromJson(request("GET", "/portfolio/overview", null));
        } catch (Exception ex) {
            throw new ApiException("加载账户汇总失败", ex);
        }
    }

    public PageResult<PortfolioHoldingBatch> listPortfolioImports(int current, int size) throws ApiException {
        try {
            JSONObject data = request("GET", "/portfolio/imports?current=" + current + "&size=" + size, null);
            PageResult<PortfolioHoldingBatch> page = new PageResult<>();
            page.total = data.optInt("total");
            page.size = data.optInt("size");
            page.current = data.optInt("current");
            JSONArray records = data.optJSONArray("records");
            if (records != null) {
                for (int i = 0; i < records.length(); i++) {
                    page.records.add(PortfolioHoldingBatch.fromJson(records.getJSONObject(i)));
                }
            }
            return page;
        } catch (Exception ex) {
            throw new ApiException("加载导入历史失败", ex);
        }
    }

    public PortfolioHoldingImportPreview previewPortfolioHoldings(List<byte[]> images) throws ApiException {
        return previewPortfolioHoldings(images, "alipay", "holding");
    }

    public PortfolioHoldingImportPreview previewPortfolioHoldings(List<byte[]> images,
                                                                   String sourceLabel) throws ApiException {
        return previewPortfolioHoldings(images, sourceLabel, "holding");
    }

    public PortfolioHoldingImportPreview previewPortfolioHoldings(List<byte[]> images,
                                                                   String sourceLabel,
                                                                   String importType) throws ApiException {
        try {
            String path = "/portfolio/imports/ocr?sourceLabel="
                    + URLEncoder.encode(sourceLabel, "UTF-8")
                    + "&importType=" + URLEncoder.encode(importType, "UTF-8");
            JSONObject data = requestMultipart(path, images);
            return PortfolioHoldingImportPreview.fromJson(data);
        } catch (Exception ex) {
            throw new ApiException("识别导入截图失败", ex);
        }
    }

    public PortfolioHoldingImportPreview portfolioHoldingImport(int importId) throws ApiException {
        try {
            return PortfolioHoldingImportPreview.fromJson(request("GET", "/portfolio/imports/" + importId, null));
        } catch (Exception ex) {
            throw new ApiException("加载导入详情失败", ex);
        }
    }

    public PortfolioHoldingConfirmResponse confirmPortfolioHoldingImport(
            int importId, PortfolioHoldingConfirmRequest requestBody) throws ApiException {
        try {
            JSONObject data = request(
                    "POST", "/portfolio/imports/" + importId + "/confirm", requestBody.toJson());
            return PortfolioHoldingConfirmResponse.fromJson(data);
        } catch (Exception ex) {
            throw new ApiException("确认入库失败", ex);
        }
    }

    private JSONObject request(String method, String path, JSONObject body) throws Exception {
        if (baseUrl == null || baseUrl.isEmpty()) {
            throw new ApiException("服务器地址无效");
        }

        URL url = new URL(baseUrl + path);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod(method);
        connection.setConnectTimeout(TIMEOUT_MS);
        connection.setReadTimeout(TIMEOUT_MS);
        connection.setRequestProperty("Accept", "application/json");
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setRequestProperty("X-Client-Source", "android");
        connection.setRequestProperty("User-Agent", "CrmMobile/Android");
        if (token != null && !token.isEmpty()) {
            connection.setRequestProperty("Authorization", "Bearer " + token);
        }

        if (body != null) {
            connection.setDoOutput(true);
            BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(connection.getOutputStream(), StandardCharsets.UTF_8));
            writer.write(body.toString());
            writer.flush();
            writer.close();
        }

        int status = connection.getResponseCode();
        InputStream stream = status >= 200 && status < 300 ? connection.getInputStream() : connection.getErrorStream();
        String text = readAll(stream);
        if (text.isEmpty()) {
            throw new ApiException("服务器未返回数据");
        }

        JSONObject response = new JSONObject(text);
        int code = response.optInt("code", -1);
        String message = response.optString("message", "请求失败 (" + status + ")");
        if (status < 200 || status >= 300 || code != 0) {
            throw new ApiException(message);
        }
        JSONObject data = response.optJSONObject("data");
        if (data == null) {
            throw new ApiException("服务器未返回数据");
        }
        return data;
    }

    private void requestNoData(String method, String path, JSONObject body) throws Exception {
        if (baseUrl == null || baseUrl.isEmpty()) {
            throw new ApiException("服务器地址无效");
        }

        URL url = new URL(baseUrl + path);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod(method);
        connection.setConnectTimeout(TIMEOUT_MS);
        connection.setReadTimeout(TIMEOUT_MS);
        connection.setRequestProperty("Accept", "application/json");
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setRequestProperty("X-Client-Source", "android");
        connection.setRequestProperty("User-Agent", "CrmMobile/Android");
        if (token != null && !token.isEmpty()) {
            connection.setRequestProperty("Authorization", "Bearer " + token);
        }
        connection.setDoOutput(true);
        BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(connection.getOutputStream(), StandardCharsets.UTF_8));
        writer.write(body.toString());
        writer.flush();
        writer.close();

        int status = connection.getResponseCode();
        InputStream stream = status >= 200 && status < 300 ? connection.getInputStream() : connection.getErrorStream();
        String text = readAll(stream);
        if (text.isEmpty()) {
            throw new ApiException("服务器未返回数据");
        }
        JSONObject response = new JSONObject(text);
        int code = response.optInt("code", -1);
        String message = response.optString("message", "请求失败 (" + status + ")");
        if (status < 200 || status >= 300 || code != 0) {
            throw new ApiException(message);
        }
    }

    private JSONObject requestMultipart(String path, List<byte[]> images) throws Exception {
        if (baseUrl == null || baseUrl.isEmpty()) {
            throw new ApiException("服务器地址无效");
        }

        String boundary = "Boundary-" + UUID.randomUUID();
        URL url = new URL(baseUrl + path);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setConnectTimeout(TIMEOUT_MS);
        connection.setReadTimeout(TIMEOUT_MS);
        connection.setDoOutput(true);
        connection.setRequestProperty("Accept", "application/json");
        connection.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);
        connection.setRequestProperty("X-Client-Source", "android");
        connection.setRequestProperty("User-Agent", "CrmMobile/Android");
        if (token != null && !token.isEmpty()) {
            connection.setRequestProperty("Authorization", "Bearer " + token);
        }

        OutputStream outputStream = connection.getOutputStream();
        for (int i = 0; i < images.size(); i++) {
            byte[] image = images.get(i);
            outputStream.write(("--" + boundary + "\r\n").getBytes(StandardCharsets.UTF_8));
            outputStream.write(("Content-Disposition: form-data; name=\"images\"; filename=\"screenshot-" + (i + 1) + ".jpg\"\r\n").getBytes(StandardCharsets.UTF_8));
            outputStream.write(("Content-Type: image/jpeg\r\n\r\n").getBytes(StandardCharsets.UTF_8));
            outputStream.write(image);
            outputStream.write("\r\n".getBytes(StandardCharsets.UTF_8));
        }
        outputStream.write(("--" + boundary + "--\r\n").getBytes(StandardCharsets.UTF_8));
        outputStream.flush();
        outputStream.close();

        int status = connection.getResponseCode();
        InputStream stream = status >= 200 && status < 300 ? connection.getInputStream() : connection.getErrorStream();
        String text = readAll(stream);
        if (text.isEmpty()) {
            throw new ApiException("服务器未返回数据");
        }

        JSONObject response = new JSONObject(text);
        int code = response.optInt("code", -1);
        String message = response.optString("message", "请求失败 (" + status + ")");
        if (status < 200 || status >= 300 || code != 0) {
            throw new ApiException(message);
        }
        JSONObject data = response.optJSONObject("data");
        if (data == null) {
            throw new ApiException("服务器未返回数据");
        }
        return data;
    }

    private String readAll(InputStream stream) throws Exception {
        if (stream == null) {
            return "";
        }
        BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8));
        StringBuilder builder = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            builder.append(line);
        }
        reader.close();
        return builder.toString();
    }
}

File: codex/crm/ios/CrmMobile/CrmMobile/ApiClient.swift
import Foundation

final class ApiClient {
    var baseURL: String
    var token: String?

    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(baseURL: String, token: String? = nil, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    func login(username: String, password: String) async throws -> LoginResult {
        let body = LoginRequest(username: username, password: password)
        return try await request(path: "/auth/login", method: "POST", body: body)
    }

    func listCustomers(current: Int, size: Int, keyword: String?) async throws -> PageResult<Customer> {
        var queryItems = [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size))
        ]
        if let keyword, !keyword.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            queryItems.append(URLQueryItem(name: "keyword", value: keyword))
        }
        return try await request(path: "/customers", queryItems: queryItems)
    }

    func customerDetail(id: Int) async throws -> Customer {
        try await request(path: "/customers/\(id)")
    }

    func listFunds(current: Int, size: Int, keyword: String?, fundType: String? = nil,
                   canBuy: Bool? = nil, sortField: String? = nil,
                   sortOrder: String? = nil) async throws -> PageResult<Fund> {
        var queryItems = [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size))
        ]
        if let keyword, !keyword.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            queryItems.append(URLQueryItem(name: "keyword", value: keyword))
        }
        if let fundType, !fundType.isEmpty {
            queryItems.append(URLQueryItem(name: "fundType", value: fundType))
        }
        if let canBuy {
            queryItems.append(URLQueryItem(name: "canBuy", value: String(canBuy)))
        }
        if let sortField { queryItems.append(URLQueryItem(name: "sortField", value: sortField)) }
        if let sortOrder { queryItems.append(URLQueryItem(name: "sortOrder", value: sortOrder)) }
        return try await request(path: "/funds", queryItems: queryItems)
    }

    func fundDetail(fundCode: String) async throws -> FundDetail {
        try await request(path: "/funds/\(fundCode)")
    }

    func listFundNavs(fundCode: String, current: Int, size: Int) async throws -> PageResult<FundNav> {
        let queryItems = [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size))
        ]
        return try await request(path: "/funds/\(fundCode)/navs", queryItems: queryItems)
    }

    func listFinanceNews(current: Int, size: Int, categoryTag: Int? = nil) async throws -> PageResult<FinanceNews> {
        var queryItems = [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size))
        ]
        if let categoryTag { queryItems.append(URLQueryItem(name: "categoryTag", value: String(categoryTag))) }
        return try await request(path: "/news", queryItems: queryItems)
    }

    func listPortfolioHoldings(current: Int, size: Int, keyword: String? = nil,
                               scope: String = "raw", sortField: String = "holdingAmount",
                               sortOrder: String = "desc") async throws -> PageResult<UserFundHolding> {
        var queryItems = [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size)),
            URLQueryItem(name: "scope", value: scope),
            URLQueryItem(name: "sortField", value: sortField),
            URLQueryItem(name: "sortOrder", value: sortOrder)
        ]
        if let keyword, !keyword.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            queryItems.append(URLQueryItem(name: "keyword", value: keyword))
        }
        return try await request(path: "/portfolio/holdings", queryItems: queryItems)
    }

    func portfolioOverview() async throws -> PortfolioOverview {
        try await request(path: "/portfolio/overview")
    }

    func listPortfolioImports(current: Int, size: Int) async throws -> PageResult<PortfolioHoldingBatch> {
        try await request(path: "/portfolio/imports", queryItems: [
            URLQueryItem(name: "current", value: String(current)),
            URLQueryItem(name: "size", value: String(size))
        ])
    }

    func previewPortfolioHoldings(
        images: [Data],
        sourceLabel: String = "alipay",
        importType: String = "holding"
    ) async throws -> PortfolioHoldingImportPreview {
        let imageParts = images.enumerated().map { index, data in
            MultipartBody.Part(name: "images", filename: "screenshot-\(index + 1).jpg", mimeType: "image/jpeg", data: data)
        }
        return try await performMultipartRequest(
            path: "/portfolio/imports/ocr?sourceLabel=\(sourceLabel)&importType=\(importType)",
            multipart: MultipartBody(parts: imageParts)
        )
    }

    func portfolioHoldingImport(importId: Int) async throws -> PortfolioHoldingImportPreview {
        try await request(path: "/portfolio/imports/\(importId)")
    }

    func confirmPortfolioHoldingImport(
        importId: Int,
        request: PortfolioHoldingConfirmRequest
    ) async throws -> PortfolioHoldingConfirmResponse {
        try await performRequest(
            path: "/portfolio/imports/\(importId)/confirm",
            method: "POST",
            queryItems: [],
            httpBody: try encoder.encode(request)
        )
    }

    func listStocks(current: Int, size: Int, keyword: String = "", sortField: String? = nil, sortOrder: String? = nil) async throws -> PageResult<StockQuote> {
        var items = [URLQueryItem(name: "current", value: String(current)), URLQueryItem(name: "size", value: String(size))]
        if !keyword.isEmpty { items.append(URLQueryItem(name: "keyword", value: keyword)) }
        if let sortField { items.append(URLQueryItem(name: "sortField", value: sortField)) }
        if let sortOrder { items.append(URLQueryItem(name: "sortOrder", value: sortOrder)) }
        return try await request(path: "/stocks", queryItems: items)
    }

    func stockHistory(stockCode: String, current: Int = 1, size: Int = 100) async throws -> PageResult<StockQuote> {
        try await request(path: "/stocks/\(stockCode)/history", queryItems: [
            URLQueryItem(name: "current", value: String(current)), URLQueryItem(name: "size", value: String(size))
        ])
    }

    private func request<T: Decodable>(
        path: String,
        method: String = "GET",
        queryItems: [URLQueryItem] = []
    ) async throws -> T {
        try await performRequest(path: path, method: method, queryItems: queryItems, httpBody: nil)
    }

    private func request<T: Decodable, B: Encodable>(
        path: String,
        method: String = "GET",
        queryItems: [URLQueryItem] = [],
        body: B
    ) async throws -> T {
        try await performRequest(path: path, method: method, queryItems: queryItems, httpBody: try encoder.encode(body))
    }

    private func performRequest<T: Decodable>(
        path: String,
        method: String,
        queryItems: [URLQueryItem],
        httpBody: Data?
    ) async throws -> T {
        var components = try components(path: path)
        if !queryItems.isEmpty {
            components.queryItems = queryItems
        }
        guard let url = components.url else {
            throw ApiError.invalidBaseURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("ios", forHTTPHeaderField: "X-Client-Source")
        request.setValue("CrmMobile/iOS", forHTTPHeaderField: "User-Agent")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = httpBody

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }
        let apiResponse = try decoder.decode(ApiResponse<T>.self, from: data)
        guard (200..<300).contains(httpResponse.statusCode), apiResponse.code == 0 else {
            throw ApiError.server(apiResponse.message)
        }
        guard let payload = apiResponse.data else {
            throw ApiError.missingData
        }
        return payload
    }

    private func performRequestNoData(path: String, method: String, httpBody: Data?) async throws {
        let components = try components(path: path)
        guard let url = components.url else {
            throw ApiError.invalidBaseURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("ios", forHTTPHeaderField: "X-Client-Source")
        request.setValue("CrmMobile/iOS", forHTTPHeaderField: "User-Agent")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = httpBody

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }
        let apiResponse = try decoder.decode(ApiResponse<EmptyResponse>.self, from: data)
        guard (200..<300).contains(httpResponse.statusCode), apiResponse.code == 0 else {
            throw ApiError.server(apiResponse.message)
        }
    }

    private func performMultipartRequest<T: Decodable>(path: String, multipart: MultipartBody) async throws -> T {
        let components = try components(path: path)
        guard let url = components.url else {
            throw ApiError.invalidBaseURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("ios", forHTTPHeaderField: "X-Client-Source")
        request.setValue("CrmMobile/iOS", forHTTPHeaderField: "User-Agent")
        request.setValue("multipart/form-data; boundary=\(multipart.boundary)", forHTTPHeaderField: "Content-Type")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = multipart.body

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }
        let apiResponse = try decoder.decode(ApiResponse<T>.self, from: data)
        guard (200..<300).contains(httpResponse.statusCode), apiResponse.code == 0 else {
            throw ApiError.server(apiResponse.message)
        }
        guard let payload = apiResponse.data else {
            throw ApiError.missingData
        }
        return payload
    }

    private func components(path: String) throws -> URLComponents {
        let normalizedBase = baseURL.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !normalizedBase.isEmpty, var components = URLComponents(string: normalizedBase + path) else {
            throw ApiError.invalidBaseURL
        }
        if components.scheme == nil {
            components = URLComponents(string: "http://" + normalizedBase + path) ?? components
        }
        return components
    }
}

private struct LoginRequest: Encodable {
    let username: String
    let password: String
}

struct PortfolioHoldingConfirmRequest: Encodable {
    let screenshotDate: String?
    let items: [PortfolioHoldingConfirmItemRequest]?
    let tradeMappings: [PortfolioTradeMappingRequest]?
}

struct PortfolioTradeMappingRequest: Encodable {
    let groupKey: String
    let fundCode: String
}

struct PortfolioHoldingConfirmItemRequest: Encodable {
    let rowNo: Int
    let fundCode: String
    let fundName: String
    let holdingAmount: Decimal?
    let holdingProfit: Decimal?
    let holdingReturnRate: Decimal?
    let holdingCost: Decimal?
    let yesterdayProfit: Decimal?
    let todayProfit: Decimal?
    let holdingShares: Decimal?
    let costNav: Decimal?
    let screenshotDate: String?
    let confidence: Decimal?
    let rawTexts: [String]
}

private struct MultipartBody {
    struct Part {
        let name: String
        let filename: String
        let mimeType: String
        let data: Data
    }

    let boundary = "Boundary-\(UUID().uuidString)"
    let body: Data

    init(parts: [Part]) {
        var data = Data()
        for part in parts {
            data.append("--\(boundary)\r\n")
            data.append("Content-Disposition: form-data; name=\"\(part.name)\"; filename=\"\(part.filename)\"\r\n")
            data.append("Content-Type: \(part.mimeType)\r\n\r\n")
            data.append(part.data)
            data.append("\r\n")
        }
        data.append("--\(boundary)--\r\n")
        self.body = data
    }
}

private extension Data {
    mutating func append(_ string: String) {
        if let value = string.data(using: .utf8) {
            append(value)
        }
    }
}
